source/internal/interactive_session.cpp

```diff
@@ -10,8 +10,11 @@ typedef std::function<void(rapidjson::Document::AllocatorType& allocator, rapidj
 
 int create_method_json(interactive_session_internal& session, const std::string& method, on_get_params getParams, bool discard, unsigned int* id, std::shared_ptr<rapidjson::Document>& methodDoc)
 {
-	std::shared_ptr<rapidjson::Document> doc(std::make_shared<rapidjson::Document>());
-	doc->SetObject();
+	std::shared_ptr<rapidjson::Document> doc(nullptr == methodDoc ? std::make_shared<rapidjson::Document>() : methodDoc);

```

Comment:
The reason this was not written this way initially was to preserve output parameter state until we are ready to return. If an error state occurs you would not want to have modified methodDoc. This seems okay since there are no error states below but would have to be rewritten if there were (unlikely). However I'm more interested in the reason for the change, is there a need to re-use the incoming pointer? It would seem that you would always expect a clean json object as well but if you pass in a non-null object this would add to it, somewhat violating the 'create_method_json' name for the function.

Reply:
Good points. Two reasons:
 - Efficiency on the calling thread (already have the document object on hand)
 - I was working towards having the [batch begin function](https://github.com/mixer/interactive-cpp/pull/80/files/2337d46dacb92bc2621421564b7f6a5988c80a61#diff-426a8659af90d0fe57bec68619f83941R20) be able to customize the document before `getParams` is called so that we don't have to [keep state in the batch entry](https://github.com/mixer/interactive-cpp/pull/80/files/2337d46dacb92bc2621421564b7f6a5988c80a61#diff-61bcf13619b1dd88b9f51026cd9f50f3R24) in the future.

Also is there a better way to get the rapidjson allocator than creating a document? Would there ever be concern with creating the json objects with an allocator from document A then using in document B?

Could always split this call into another function.

Reply:
The allocator is bound to the document, definitely can't mix and match them. rapidjson has functions to move objects around without copies which might solve this.